File: apps/android/core/model/src/main/kotlin/app/gauja/core/model/links/DeepLink.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.model.links

import app.gauja.core.model.media.TmdbId
import app.gauja.core.model.servers.ProfileId
import app.gauja.core.model.servers.ServerProfile
import java.net.URI
import java.util.UUID

sealed interface LinkTarget {
    data class Media(val kind: Kind, val id: TmdbId) : LinkTarget

    data object Requests : LinkTarget

    data class Issue(val id: Int) : LinkTarget

    data class ResetPassword(val guid: UUID) : LinkTarget

    enum class Kind {
        MOVIE,
        TV,
        PERSON,
        COLLECTION,
    }
}

data class DeepLink(val profileIds: List<ProfileId>, val target: LinkTarget)

fun parseDeepLink(raw: String, profiles: List<ServerProfile>): DeepLink? {
    val url =
        try {
            URI(raw)
        } catch (_: java.net.URISyntaxException) {
            return null
        }
    if (url.userInfo != null || url.query != null || url.fragment != null) return null
    val path = url.rawPath ?: return null
    if (path.contains("%") || path.contains("\\")) return null
    val candidates: List<ProfileId>
    val route: String
    if (url.scheme == "gauja" && url.host == "server" && url.port == -1) {
        val parts = path.trimStart('/').split('/', limit = 2)
        if (parts.size != 2) return null
        val id =
            try {
                ProfileId(canonicalUuid(parts[0]) ?: return null)
            } catch (_: IllegalArgumentException) {
                return null
            }
        candidates = profiles.filter { it.id == id }.map { it.id }
        route = parts[1]
    } else {
        val matching =
            profiles.filter {
                val base = URI(it.address.value)
                url.scheme == base.scheme &&
                    url.host == base.host &&
                    url.port == base.port &&
                    path.startsWith(base.path + "/")
            }
        if (matching.isEmpty()) return null
        val prefix = matching.maxOf { URI(it.address.value).path.length }
        candidates = matching.filter { URI(it.address.value).path.length == prefix }.map { it.id }
        route = path.drop(prefix + 1)
    }
    if (candidates.isEmpty()) return null
    val parts = route.split('/')
    val target =
        if (parts == listOf("requests")) LinkTarget.Requests
        else {
            if (parts.size != 2) return null
            when (parts[0]) {
                "resetpassword",
                "reset-password" ->
                    try {
                        LinkTarget.ResetPassword(canonicalUuid(parts[1]) ?: return null)
                    } catch (_: IllegalArgumentException) {
                        return null
                    }
                "issues" ->
                    LinkTarget.Issue(parts[1].toIntOrNull()?.takeIf { it > 0 } ?: return null)
                else -> {
                    val kind =
                        LinkTarget.Kind.entries.firstOrNull { it.name.lowercase() == parts[0] }
                            ?: return null
                    LinkTarget.Media(
                        kind,
                        TmdbId(parts[1].toIntOrNull()?.takeIf { it > 0 } ?: return null),
                    )
                }
            }
        }
    return DeepLink(candidates, target)
}

private fun canonicalUuid(value: String): UUID? =
    UUID.fromString(value).takeIf { it.toString().equals(value, ignoreCase = true) }

File: apps/ios/Packages/Model/Sources/Model/links/DeepLink.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

public enum LinkTarget: Sendable, Equatable {
    public enum Kind: String, Sendable { case movie, tv, person, collection }
    case media(Kind, TMDBID)
    case requests
    case issue(Int)
    case resetPassword(UUID)
}
public struct DeepLink: Sendable, Equatable {
    public let profileIDs: [ProfileID]
    public let target: LinkTarget
}

public func parseDeepLink(_ raw: String, profiles: [ServerProfile]) -> DeepLink? {
    guard let url = URLComponents(string: raw), url.user == nil, url.password == nil, url.query == nil,
        url.fragment == nil,
        !url.percentEncodedPath.contains("%"), !url.path.contains("\\")
    else { return nil }
    let candidates: [ProfileID]
    let route: String
    if url.scheme == "gauja", url.host == "server", url.port == nil {
        let parts = url.path.dropFirst().split(separator: "/", maxSplits: 1, omittingEmptySubsequences: false)
        guard parts.count == 2, let uuid = UUID(uuidString: String(parts[0])) else { return nil }
        candidates = profiles.filter { $0.id.rawValue == uuid }.map(\.id)
        route = String(parts[1])
    } else {
        let matches = profiles.filter {
            url.scheme == $0.address.url.scheme && url.host == $0.address.url.host && url.port == $0.address.url.port
                && url.path.hasPrefix($0.address.url.path + "/")
        }
        guard let prefix = matches.map({ $0.address.url.path.count }).max() else { return nil }
        candidates = matches.filter { $0.address.url.path.count == prefix }.map(\.id)
        route = String(url.path.dropFirst(prefix + 1))
    }
    guard !candidates.isEmpty else { return nil }
    let parts = route.components(separatedBy: "/")
    let target: LinkTarget
    if parts == ["requests"] {
        target = .requests
    } else {
        guard parts.count == 2 else { return nil }
        switch parts[0] {
        case "resetpassword", "reset-password":
            guard let uuid = UUID(uuidString: parts[1]) else { return nil }
            target = .resetPassword(uuid)
        case "issues":
            guard let id = Int(parts[1]), id > 0 else { return nil }
            target = .issue(id)
        default:
            guard let kind = LinkTarget.Kind(rawValue: parts[0]), let rawID = Int(parts[1]),
                let id = TMDBID(rawValue: rawID)
            else { return nil }
            target = .media(kind, id)
        }
    }
    return DeepLink(profileIDs: candidates, target: target)
}
